platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/configurationcache/EncryptionService.kt

```diff
@@ -251,9 +239,91 @@ class KeyStoreKeySource(
         return newKey
     }
 
+    private
+    fun cacheBuilderFor(): CacheBuilder =
+        cacheBuilderFactory
+            .run { customKeyStoreDir?.let { createCacheBuilderFactory(it) } ?: this }
+            .createCacheBuilder("cc-keystore")
+            .withDisplayName("Gradle Configuration Cache keystore")
+
+    override fun getKey(): SecretKey {
+        return cacheBuilderFor()
+            .withInitializer {
+                createKeyStoreAndGenerateKey(keyStoreFile())
+            }.open().useToRun {
+                try {
+                    loadSecretKeyFromExistingKeystore(keyStoreFile())
+                } catch (loadException: Exception) {
+                    // try to recover from a tampered-with keystore by generating a new key
+                    // TODO:configuration-cache do we really need this?
+                    try {
+                        createKeyStoreAndGenerateKey(keyStoreFile())
+                    } catch (e: Exception) {
+                        e.addSuppressed(loadException)
+                        throw e
+                    }
+                }
+            }
+    }
+
+    private
+    fun PersistentCache.keyStoreFile(): File =
+        File(baseDir, "gradle.keystore")
+
     companion object {
         // JKS does not support non-PrivateKeys
         const val KEYSTORE_TYPE = "pkcs12"
         val KEYSTORE_PASSWORD = charArrayOf('c', 'c')
     }
 }
+
+
+class EnvironmentVarKeySource(encryptionAlgorithm: String) : SecretKeySource {
+    private
+    val secretKey: SecretKey by lazy {
+        Base64.getDecoder().decode(getKeyAsBase64()).let { keyAsBytes ->
+            SecretKeySpec(keyAsBytes, encryptionAlgorithm)
+        }
+    }
+
+    private
+    fun getKeyAsBase64(): String = System.getenv(GRADLE_ENCRYPTION_KEY_ENV_KEY) ?: ""
+
+    override fun getKey(): SecretKey = secretKey
+
+    override val sourceDescription: String
+        get() = "$GRADLE_ENCRYPTION_KEY_ENV_KEY environment variable"
+
+    companion object {
+        const val GRADLE_ENCRYPTION_KEY_ENV_KEY = "GRADLE_ENCRYPTION_KEY"
```

Comment:
Unless there are plans to expand this key's use beyond the configuration cache it might make sense to name it `GRADLE_CONFIGURATION_CACHE_ENCRYPTION_KEY` as that gives a bit more context. Could shorten it to `GRADLE_CONFIG_CACHE_SECRET` if that's preferable.